File: packages/exochain-sdk/src/authority/chain.ts
/**
 * Authority chain builder and validator.
 *
 * An authority chain is an ordered list of delegation links where each
 * `grantee` is the `grantor` of the next link. The chain terminates at a
 * specific actor. Mirrors the Rust SDK's `AuthorityChainBuilder` API.
 */

import { AuthorityError } from '../errors.js';
import type { Did } from '../types.js';
import { validateDid } from '../identity/did.js';

/** A single delegation link. */
export interface ChainLink {
  readonly grantor: Did;
  readonly grantee: Did;
  readonly permissions: readonly string[];
}

/** A fully validated authority chain. */
export interface ValidatedChain {
  readonly depth: number;
  readonly links: readonly ChainLink[];
  readonly terminal: Did;
}

/** Fluent builder for an authority chain. */
export class AuthorityChainBuilder {
  readonly #links: ChainLink[] = [];

  /** Append a delegation link. Returns `this` for chaining. */
  public addLink(
    grantor: Did | string,
    grantee: Did | string,
    permissions: readonly string[],
  ): this {
    const grantorDid = typeof grantor === 'string' ? validateDid(grantor) : grantor;
    const granteeDid = typeof grantee === 'string' ? validateDid(grantee) : grantee;
    this.#links.push({
      grantor: grantorDid,
      grantee: granteeDid,
      permissions: [...permissions],
    });
    return this;
  }

  /**
   * Validate the chain topology and return a {@link ValidatedChain}.
   *
   * Rules:
   * - At least one link.
   * - For each consecutive pair, `links[i].grantee === links[i+1].grantor`.
   * - The final `grantee` must equal `terminalActor`.
   */
  public build(terminalActor: Did | string): ValidatedChain {
    const terminal =
      typeof terminalActor === 'string' ? validateDid(terminalActor) : terminalActor;

    if (this.#links.length === 0) {
      throw new AuthorityError('authority chain is empty');
    }

    for (let i = 0; i < this.#links.length - 1; i++) {
      const a = this.#links[i];
      const b = this.#links[i + 1];
      if (a === undefined || b === undefined) continue;
      if (a.grantee !== b.grantor) {
        throw new AuthorityError(
          `broken delegation at index ${i}: ${a.grantee} !== ${b.grantor}`,
        );
      }
    }

    const last = this.#links[this.#links.length - 1];
    if (last === undefined) {
      throw new AuthorityError('authority chain is empty');
    }
    if (last.grantee !== terminal) {
      throw new AuthorityError(
        `terminal mismatch: chain ends at ${last.grantee} but terminal is ${terminal}`,
      );
    }

    return {
      depth: this.#links.length,
      links: [...this.#links],
      terminal,
    };
  }
}

File: packages/exochain-py/exochain/authority/chain.py
"""Authority chain builder and validated chain type.

An *authority chain* is an ordered list of delegation links where the grantee
of each link is the grantor of the next. The chain terminates at a specific
terminal actor. :class:`AuthorityChainBuilder` accumulates links and performs
topology validation in :meth:`AuthorityChainBuilder.build`.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict

from ..errors import AuthorityError
from ..types import Did


class ChainLink(BaseModel):
    """A single delegation link: grantor → grantee with scoped permissions."""

    model_config = ConfigDict(frozen=True)

    grantor: Did
    grantee: Did
    permissions: list[str]


class ValidatedChain(BaseModel):
    """An authority chain that has passed topology validation."""

    model_config = ConfigDict(frozen=True)

    depth: int
    links: list[ChainLink]
    terminal: Did


class AuthorityChainBuilder:
    """Fluent builder that accumulates links and validates on ``build``."""

    def __init__(self) -> None:
        self._links: list[ChainLink] = []

    def add_link(
        self, grantor: Did, grantee: Did, permissions: list[str]
    ) -> AuthorityChainBuilder:
        """Append a delegation link to the chain."""
        if not isinstance(permissions, list) or not all(
            isinstance(p, str) for p in permissions
        ):
            raise AuthorityError("permissions must be a list of strings")
        self._links.append(
            ChainLink(grantor=grantor, grantee=grantee, permissions=list(permissions))
        )
        return self

    def build(self, terminal_actor: Did) -> ValidatedChain:
        """Validate the accumulated links and produce a :class:`ValidatedChain`.

        Validation rules:
          * The chain must contain at least one link.
          * For each consecutive pair, ``links[i].grantee == links[i+1].grantor``.
          * ``links[-1].grantee == terminal_actor``.

        Raises:
            AuthorityError: if any validation rule is violated.
        """
        if not self._links:
            raise AuthorityError("authority chain is empty")

        for a, b in zip(self._links, self._links[1:], strict=False):
            if a.grantee != b.grantor:
                raise AuthorityError(
                    f"broken delegation: {a.grantee} != {b.grantor}"
                )

        last = self._links[-1]
        if last.grantee != terminal_actor:
            raise AuthorityError(
                f"terminal mismatch: chain ends at {last.grantee} "
                f"but terminal_actor is {terminal_actor}"
            )

        return ValidatedChain(
            depth=len(self._links),
            links=list(self._links),
            terminal=terminal_actor,
        )


__all__ = ["AuthorityChainBuilder", "ChainLink", "ValidatedChain"]
